### PR TITLE
8189685: need PerfMemory class update and a volatile_static_field support in VMStructs

### PR DESCRIPTION
**Pull request by plummercj**
During [JDK-8151815](https://bugs.openjdk.org/browse/JDK-8151815) it was noted that the PerfMemory _initialized and _destroyed fields should be volatile, but VMStructs didn't have the needed support for doing that, so it was left as a future task. @YaSuenag provided a patch at the time to take care of the VMStructs support. I've integrated it, although it was far from clean due to some changes in VMStructs, and also moving OrderAccess::release_store to Atomic::release_store.

One other change I made to the patch had to do with consistency with using "volatile static" vs "static volatile". We already have volatile_nonstatic_field. The patch renamed static_ptr_volatile_field to static_volatile_field to make it more general purpose, but this was inconsistent with the name of volatile_nonstatic_field, so I chose the name volatile_static_field instead. This carried over into some other areas like the names of the GENERATE_VOLATILE_STATIC_VM_STRUCT_ENTRY and CHECK_VOLATILE_STATIC_VM_STRUCT_ENTRY macros.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8189685](https://bugs.openjdk.org/browse/JDK-8189685): need PerfMemory class update and a volatile_static_field support in VMStructs (**Bug** - P4)

### Reviewers
 * [Yasumasa Suenaga](https://openjdk.org/census#ysuenaga) (@YaSuenag - **Reviewer**)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15373/head:pull/15373` \
`$ git checkout pull/15373`

Update a local copy of the PR: \
`$ git checkout pull/15373` \
`$ git pull https://git.openjdk.org/jdk.git pull/15373/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15373`

View PR using the GUI difftool: \
`$ git pr show -t 15373`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15373.diff">https://git.openjdk.org/jdk/pull/15373.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15373#issuecomment-1687095922)

### CODE REVIEW

**Review by YaSuenag — APPROVED** [review]
LGTM.
Thank you for heading up for this issue!

File: src/hotspot/share/runtime/perfMemory.cpp [hunk 1]
```diff
@@ -197,7 +197,7 @@ void PerfMemory::destroy() {
     delete_memory_region();
   }
 
-  _destroyed = true;
+  Atomic::release_store(&_destroyed, 1);
```
[the remark this answers is not in the document]
Comment (plummercj) [reply src/hotspot/share/runtime/perfMemory.cpp]: Ok. I've updated it.

File: src/hotspot/share/runtime/perfMemory.hpp [hunk 2]
```diff
@@ -134,7 +134,7 @@ class PerfMemory : AllStatic {
     static size_t used() { return (size_t) (_top - _start); }
     static size_t capacity() { return _capacity; }
     static bool is_initialized();
-    static bool is_destroyed() { return _destroyed; }
+    static bool is_destroyed();
```
Comment (YaSuenag) [inline src/hotspot/share/runtime/perfMemory.hpp:137]: So don't we need to make changes here? I think we can keep current implementation ( `is_destroyed()` )

File: src/hotspot/share/runtime/perfMemory.cpp [hunk 3]
```diff
@@ -51,9 +51,9 @@ char*                    PerfMemory::_start = nullptr;
 char*                    PerfMemory::_end = nullptr;
 char*                    PerfMemory::_top = nullptr;
 size_t                   PerfMemory::_capacity = 0;
-int                      PerfMemory::_initialized = false;
+volatile int             PerfMemory::_initialized = 0;
 PerfDataPrologue*        PerfMemory::_prologue = nullptr;
-bool                     PerfMemory::_destroyed = false;
+volatile int             PerfMemory::_destroyed = 0;
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/perfMemory.cpp:56]: Why was this changed to int?

File: src/hotspot/share/runtime/perfMemory.hpp (hunk 2 above)
Comment (plummercj) [reply src/hotspot/share/runtime/perfMemory.hpp:137]: I was also partly going for consistency with _initialized and _destroyed w.r.t. int vs bool and 0/1 vs false/true. I can go back to making _destroyed all bool.

Comment (plummercj) [reply src/hotspot/share/runtime/perfMemory.hpp]: Update done. I should also point out that _destroyed is not exported from VMStructs, so could have been made volatile long ago without needing to add any VMStructs support. SA only looks at _initialized, but I suppose you could argue that it also should be checking _destroyed before accessing PerfMemory.

Comment (YaSuenag) [reply src/hotspot/share/runtime/perfMemory.hpp]: > SA only looks at _initialized, but I suppose you could argue that it also should be checking _destroyed before accessing PerfMemory.

In JDK-8151815, we do not clear fields in `PerfMemory` because they are used in SA even if they are freed, so we can ignore `_destroyed` in SA. We've discussed about this in [review thread of JDK-8151815](https://mail.openjdk.org/pipermail/serviceability-dev/2017-October/022050.html).

File: src/hotspot/share/runtime/vmStructs.hpp [hunk 4]
```diff
@@ -207,7 +207,7 @@ class VMStructs {
 
 // This macro checks the type of a static pointer volatile VMStructEntry by comparing pointer types,
 // e.g.: "static ObjectMonitor * volatile g_block_list;"
-#define CHECK_STATIC_PTR_VOLATILE_VM_STRUCT_ENTRY(typeName, fieldName, type)       \
+#define CHECK_VOLATILE_STATIC_VM_STRUCT_ENTRY(typeName, fieldName, type)       \
  {type volatile * dummy = &typeName::fieldName; }
```
Comment (sspitsyn) [inline src/hotspot/share/runtime/vmStructs.hpp:211]: It is not clear why the `PTR_` suffix is removed from the name.

Comment (dholmes-ora) [reply src/hotspot/share/runtime/vmStructs.hpp:211]: The intent was to generalize it but the comments also need changing, and now I'm not clear what this is actually doing.

Comment (plummercj) [reply src/hotspot/share/runtime/vmStructs.hpp:211]: This change wasn't in the original patch, but the `GENERATE_STATIC_PTR_VOLATILE_VM_STRUCT_ENTRY` change above it was.  I found it when I noticed the use of `GENERATE_STATIC_PTR_VOLATILE_VM_STRUCT_ENTRY` and thought it should be renamed too. I was also doing "static volatile" -> "volatile stack" at the time so that is also part of this macros rename. 

As David points out, the purpose is to make it more general purpose, although for this macro only the name change is needed (`GENERATE_STATIC_PTR_VOLATILE_VM_STRUCT_ENTRY also has a minor implementation change that was needed).

The comments for both of these macros references "static pointer volatile" entries. That should changed to "volatile static". The type no longer needs to be a pointer type. 

I'm not certain what this CHECK macro does, but it was pre-existing, and there is also the `CHECK_STATIC_VM_STRUCT_ENTRY` macro above it which is identical except of the absence of "volatile" in the name and implementation. I think it verifies that the type  specified in the vmstructs definition of the field is actually the same as the field itself.  The macro will generate a compiler error if a cast was needed. I think this type of check can only be done with static fields. You would need an instance of the class/struct to do a similar check on an instance field. Thus CHECK_NONSTATIC_VM_STRUCT_ENTRY is much more complex and seems to be a runtime check that will assert if it fails.

Comment (dholmes-ora) [reply src/hotspot/share/runtime/vmStructs.hpp:211]: I can't convince myself that `volatile` is in the right place in `type volatile * dummy`. But `volatile * dummy` means that the pointer is volatile as opposed to the type being pointed to, which would match with the original macro name's use of `PTR_VOLATILE`.

Comment (plummercj) [reply src/hotspot/share/runtime/vmStructs.hpp:211]: But in that case the use of the macro doesn't make much sense. Regardless of the field type, vmstructs doesn't deal with volatile pointers to fields. It deals with volatile fields (that can be pointers or scalar). I think if anything the macro has always been wrong and should contain `volatile type * dummy`. It probably never fails because it is taking a non-volatile type and assigning it to a volatile type, which always works if the types are otherwise the same.

Comment (plummercj) [reply src/hotspot/share/runtime/vmStructs.hpp:211]: Well of course that change blows up

invalid conversion from 'ClassLoaderData* volatile*' to 'volatile ClassLoaderData**'

There's something I'm not understanding here. I use to understand the volatile ordering syntax and how to read it properly, but that was long long ago.

Comment (plummercj) [reply src/hotspot/share/runtime/vmStructs.hpp:211]: I take back the following statement:

> It probably never fails because it is taking a non-volatile type and assigning it to a volatile type, which always works if the types are otherwise the same.

Since the field type is volatile, the current approach must be correct. Otherwise there would be cast errors if the LHS did not have volatile in the right place.

Comment (plummercj) [reply src/hotspot/share/runtime/vmStructs.hpp:211]: @dholmes-ora I'm not sure what is needed to move forward with this. Are you OK with the change?

**Review by dholmes-ora — APPROVED** [review]
I'm not quite certain about the intent with the old naming for the macros, but the new names certainly don't hurt anything. So this seems okay.

Thanks.

**Review by sspitsyn — APPROVED** [review]
Looks good.
Thanks,
Serguei

**Review by YaSuenag — APPROVED** [review]
I agree with this change.

**Comment by plummercj** [comment]
Thanks for the reviews Yasumasa, David, and Serguei!

/integrate
